File: Algorithm/Python/064. Minimum Path Sum.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-05-10 09:06:36
# @Last modified by:   WuLC
# @Last Modified time: 2016-05-10 09:06:56
# @Email: liangchaowu5@gmail.com

# DP
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])
        for i in xrange(m):
            for j in xrange(n):
                if i == 0 and j == 0:
                    continue
                elif i == 0 :
                    grid[i][j] += grid[i][j-1]
                elif j == 0 :
                    grid[i][j] += grid[i-1][j]
                else:
                    grid[i][j] = min(grid[i][j]+grid[i][j-1],grid[i][j]+grid[i-1][j])
        return grid[m-1][n-1]
        

File: Algorithm/Java/064. Minimum Path Sum.java
/**
* Author: WuLC
* Date:   2017-01-08 17:10:17
* Last modified by:   WuLC
* Last Modified time: 2017-01-08 17:10:42
* Email: liangchaowu5@gmail.com
*/

// two dimensional dp
public class Solution 
{
    public int minPathSum(int[][] grid) 
    {
        int m = grid.length, n = grid[0].length;
        if (m == 0 || n == 0) return 0;
        int[][] dp = new int[m][n];
        for(int i=0; i<m; i++)
        {
            for(int j=0; j<n; j++)
            {
                if(i==0 && j==0) dp[i][j] = grid[i][j];
                else if(i==0) dp[i][j] = dp[i][j-1] + grid[i][j];
                else if(j==0) dp[i][j] = dp[i-1][j] + grid[i][j];
                else dp[i][j] = Math.min(dp[i-1][j], dp[i][j-1]) + grid[i][j];
            }
        }
        return dp[m-1][n-1];
    }
}